### PR TITLE
GetEntity Support for ReadOnlyDB and SecondaryDB

### PR DESCRIPTION
**Pull request by jaykorean**
# Summary
`GetEntity` API support for ReadOnly DB and Secondary DB.
- Introduced `GetImpl()` with `GetImplOptions` in `db_impl_readonly` and refactored current `Get()` logic into `GetImpl()` so that look up logic can be reused for `GetEntity()` (Following the same pattern as `DBImpl::Get()` and `DBImpl::GetEntity()`)
- Introduced `GetImpl()` with `GetImplOptions` in `db_impl_secondary` and refactored current `GetImpl()` logic. This is to make `DBImplSecondary::Get/GetEntity` consistent with `DBImpl::Get/GetEntity` and `DBImplReadOnly::Get/GetEntity`
- `GetImpl()` in `db_impl` is now virtual. both `db_impl_readonly` and `db_impl_secondary`'s `Get()` override are no longer needed since all three dbs now have the same `Get()` which calls `GetImpl()` internally.
- `GetImpl()` in `DBImplReadOnly` and `DBImplSecondary` now pass in `columns` instead of `nullptr` in lookup functions like `memtable->get()`
- Introduced `GetEntity()` API in `DBImplReadOnly` and `DBImplSecondary` which simply calls `GetImpl()` with `columns` set in `GetImplOptions`.
- Introduced `Env::IOActivity::kGetEntity` and set read_options.io_activity to `Env::IOActivity::kGetEntity` for `GetEntity()` operations (in db_impl)

# Test Plan
**Unit Tests**
- Added verification in `DBWideBasicTest::PutEntity` by Reopening DB as ReadOnly with the same setup.
- Added verification in `DBSecondaryTest::ReopenAsSecondary` by calling `PutEntity()` and `GetEntity()` on top of existing `Put()` and `Get()`
- `make -j64 check`

**Crash Tests**
- `python3 tools/db_crashtest.py blackbox --max_key=25000000 --write_buffer_size=4194304 --max_bytes_for_level_base=2097152 --target_file_size_base=2097152 --periodic_compaction_seconds=0 --use_put_entity_one_in=10 --use_get_entity=1 --duration=60 --inter
val=10`
- `python3 tools/db_crashtest.py blackbox --simple --max_key=25000000 --write_buffer_size=4194304 --max_bytes_for_level_base=2097152 --target_file_size_base=2097152 --periodic_compaction_seconds=0 --use_put_entity_one_in=10 --use_get_entity=1 `
- `python3 tools/db_crashtest.py blackbox --cf_consistency --max_key=25000000 --write_buffer_size=4194304 --max_bytes_for_level_base=2097152 --target_file_size_base=2097152 --periodic_compaction_seconds=0 --use_put_entity_one_in=10 --use_get_entity=1 --duration=60 --inter
val=10`

### CODE REVIEW

File: db/db_impl/db_impl_readonly.cc [hunk 1]
```diff
@@ -29,99 +29,150 @@ DBImplReadOnly::DBImplReadOnly(const DBOptions& db_options,
...
-    return Status::InvalidArgument(
-        "Can only call Get with `ReadOptions::io_activity` is "
-        "`Env::IOActivity::kUnknown` or `Env::IOActivity::kGet`");
-  }
-  ReadOptions read_options(_read_options);
-  if (read_options.io_activity == Env::IOActivity::kUnknown) {
-    read_options.io_activity = Env::IOActivity::kGet;
-  }
-  assert(pinnable_val != nullptr);
-  PERF_CPU_TIMER_GUARD(get_cpu_nanos, immutable_db_options_.clock);
```
Comment (jaykorean) [inline db/db_impl/db_impl_readonly.cc:54]: Moving timer _after_ timestamp check + clearing timestamps (line 59) to be consistent with `DbImpl::GetImpl()` - https://github.com/facebook/rocksdb/blob/137cd4bb750bec6071254f3b47e89db4d2b58593/db/db_impl/db_impl.cc#L2029-L2037

File: include/rocksdb/env.h [hunk 2]
```diff
@@ -446,6 +446,8 @@ class Env : public Customizable {
     kDBIterator = 5,
     kVerifyDBChecksum = 6,
     kVerifyFileChecksums = 7,
+    kGetEntity = 8,
+    kMultiGetEntity = 9,
```
Comment (jaykorean) [inline include/rocksdb/env.h:450]: Will address this in MultiGetEntity APIs in the later PR.

**Comment by hx235** [comment]
For IOActivity stuff, you might also want to add it to db_stress like https://github.com/facebook/rocksdb/blob/main/db_stress_tool/db_stress_test_base.cc#L1081-L1085 to verify the pass-down IOActivity in a thread aligns with the actual activity the thread is doing. See https://github.com/facebook/rocksdb/pull/11444/files for more example. 

You can then manually run some stress test on ReadOnlyDB/SecondaryDB/GetEntity paths to ensure this [assertion](https://github.com/facebook/rocksdb/blob/main/db_stress_tool/db_stress_env_wrapper.h#L30-L31) pass.

File: db/db_impl/db_impl.h [hunk 3]
```diff
@@ -643,8 +643,8 @@ class DBImpl : public DB {
   // get_impl_options.key via get_impl_options.value
   // If get_impl_options.get_value = false get merge operands associated with
   // get_impl_options.key via get_impl_options.merge_operands
-  Status GetImpl(const ReadOptions& options, const Slice& key,
-                 GetImplOptions& get_impl_options);
+  virtual Status GetImpl(const ReadOptions& options, const Slice& key,
+                         GetImplOptions& get_impl_options);
```
Comment (ltamasi) [inline db/db_impl/db_impl.h:647]: Great call 👍👍

**Review by ltamasi — APPROVED** [review]
Thanks for the PR and all the code improvements @jaykorean!!

**Pull request closed without merging**